_posts/2018-11-29-GraphQL-Stitching.markdown

```diff
@@ -0,0 +1,392 @@
+          return info.mergeInfo.delegateToSchema({
+            schema: localSchema,
+            operation: "query",
+            fieldName: "artist",
+            args: {
+              id,
+            },
+            context,
+            info,
+            transforms: (convectionSchema as any).transforms,
+          })
+        },
+      },
+    },
+  },
+})
+```
+
+This file consolidates the two steps of merging and then stitching:
+
+```diff
+export const mergeSchemas = async () => {
+  const convectionSchema = await executableConvectionSchema()
++ const convectionStitching = consignmentStitchingEnvironment(localSchema, convectionSchema)
+
+  const gravitySchema = await executableGravitySchema()
+
+  // The order should only matter in that extension schemas come after the
+  // objects that they are expected to build upon
+  const mergedSchema = _mergeSchemas({
+    schemas: [
+      gravitySchema,
+      localSchema,
+      convectionSchema,
++      convectionStitching.extensionSchema,
++    ],
++    resolvers: {
++      ...convectionStitching.resolvers,
++    },
+  })
+
+  return mergedSchema
+}
+```
+
+We extend the merge schema function to also include the SDL for our stitching, and de-structure in the extension
+resolvers. We're still exploring how to write _useful_ tests for this part.
+
+## Validating your changes
+
+We had some useful tools which were used to make the switch to using schema-stitching in production.
+
+1. **Stored Queries**
+
+   In order to validate that the runtime behavior of our queries wasn't changing, we used the [persistent
+   queries][ps] generated by our iOS app Emission to create JSON dumps of the results of many API calls in both
+   stitched and un-stitched environments in a script and compared the results.
+
+1. **SDL dump comparison**
+
+   We can use the GraphQL type system to validate our changes don't break clients. We used a schema dump script to
+   validate the type system was the same across stitched and un-stitched environments.
+
+## Alternatives
+
+Our stitching implementation used metaphysics as a local schema - e.g. the merging and the stitching occurred
+inside the same source code as our main GraphQL API. This wasn't the only way we cold achieve a single source of
+truth, I explored an alternative which is to treat all schemas as remote schemas. Basically having an API gateway
+in-front of our API gateway. I called it [incorporeal][inc], it's actually a tiny amount of code - a testament to
+modern GraphQL techniques in JavaScript.
+
+The advantage here is that we take reduce the acceleration of growing complexity in Metaphysics completely, because
+the merging and stitching occurs outside of it. Metaphysics is merged and stitched in, just as all our other APIs
+are.
+
+The downside is that it's another hop to get what you want, and changes could require being updated in more places.
+We'd be able to use the above ideas for validating that the API is working as expected.
```

Comment:
I like that we see the pros/cons of the alternative, but I'm left wondering - why did we settle on our solution over the alternative? That might be a useful thing to state.